File: 0072) Edit Distance.cpp
class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.size();
        int n = word2.size();
        vector<vector<int>> dp(m+1, vector<int>(n+1, -1));
        return helper(word1, word2, 0, 0, dp);
    }
    int helper(string &word1, string &word2, int i, int j, vector<vector<int>> &dp)
    {
        int m = word1.size();
        int n = word2.size();
        if(i == m)
        {
            return n-j;
        }
        if(j == n)
        {
            return m-i;
        }
        if(dp[i][j] != -1)
        {
            return dp[i][j];
        }
        if(word1[i] == word2[j])
        {
            dp[i][j] = helper(word1, word2, i+1, j+1, dp);
        }
        else
        {
            int insert = 1 + helper(word1, word2, i, j+1, dp);
            int remove = 1 + helper(word1, word2, i+1, j, dp);
            int replace = 1 + helper(word1, word2, i+1, j+1, dp);
            dp[i][j] = min({insert, remove, replace});
        }
        return dp[i][j];
    }
};

File: 0072) Edit Distance.java
class Solution {
    public int minDistance(String word1, String word2) {
        int m = word1.length();
        int n = word2.length();
        int[][] dp = new int[m+1][n+1];
        for(int i = 0; i < m; i++)
        {
            for(int j = 0; j < n; j++)
            {
                dp[i][j] = -1;
            }
        }
        return helper(0, 0, word1, word2, dp);
    }
    public int helper(int i, int j, String word1, String word2, int[][] dp)
    {
        int m = word1.length();
        int n = word2.length();
        if(i == m)
        {
            return n-j;
        }
        if(j == n)
        {
            return m-i;
        }
        if(dp[i][j] != -1)
        {
            return dp[i][j];
        }
        if(word1.charAt(i) == word2.charAt(j))
        {
            dp[i][j] = helper(i+1, j+1, word1, word2, dp);
        }
        else
        {
            int insert = 1 + helper(i, j+1, word1, word2, dp);
            int remove = 1 + helper(i+1, j, word1, word2, dp);
            int replace = 1 + helper(i+1, j+1, word1, word2, dp);
            dp[i][j] = Math.min(insert, Math.min(remove, replace));
        }
        return dp[i][j];
    }
}
